datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -500,302 +387,359 @@ fn optimize_join(
-                    state.filters.push((predicate, columns));
-                    Ok(())
-                })?;
-
-            optimize(filter.input(), state)
+fn push_down_join(
+    plan: &LogicalPlan,
+    join: &Join,
+    parent_predicate: Option<&Expr>,
+) -> Result<Option<LogicalPlan>> {
+    let mut predicates = match parent_predicate {
+        Some(parent_predicate) => {
+            utils::split_conjunction_owned(utils::cnf_rewrite(parent_predicate.clone()))
         }
-        LogicalPlan::Projection(Projection {
-            input,
-            expr,
-            schema,
-        }) => {
-            // A projection is filter-commutable, but re-writes all predicate expressions
-            // collect projection.
-            let projection = schema
-                .fields()
-                .iter()
-                .enumerate()
-                .flat_map(|(i, field)| {
-                    // strip alias, as they should not be part of filters
-                    let expr = match &expr[i] {
-                        Expr::Alias(expr, _) => expr.as_ref().clone(),
-                        expr => expr.clone(),
+        None => vec![],
+    };
+
+    // Convert JOIN ON predicate to Predicates
+    let on_filters = join
+        .filter
+        .as_ref()
+        .map(|e| utils::split_conjunction_owned(e.clone()))
+        .unwrap_or_else(Vec::new);
+
+    if join.join_type == JoinType::Inner {
+        // For inner joins, duplicate filters for joined columns so filters can be pushed down
+        // to both sides. Take the following query as an example:
+        //
+        // ```sql
+        // SELECT * FROM t1 JOIN t2 on t1.id = t2.uid WHERE t1.id > 1
+        // ```
+        //
+        // `t1.id > 1` predicate needs to be pushed down to t1 table scan, while
+        // `t2.uid > 1` predicate needs to be pushed down to t2 table scan.
+        //
+        // Join clauses with `Using` constraints also take advantage of this logic to make sure
+        // predicates reference the shared join columns are pushed to both sides.
+        // This logic should also been applied to conditions in JOIN ON clause
+        let join_side_filters = predicates
+            .iter()
+            .chain(on_filters.iter())
+            .filter_map(|predicate| {
+                let mut join_cols_to_replace = HashMap::new();
+                let columns = match predicate.to_columns() {
+                    Ok(columns) => columns,
+                    Err(e) => return Some(Err(e)),
+                };
+
+                for col in columns.iter() {
+                    for (l, r) in join.on.iter() {
+                        if col == l {
+                            join_cols_to_replace.insert(col, r);
+                            break;
+                        } else if col == r {
+                            join_cols_to_replace.insert(col, l);
+                            break;
+                        }
+                    }
+                }
```

Comment:
There is `break` to return the inner loop.   I think for join conditions: `on(a.id = b.id and a.id = b.id2) where b.id = 10`, we should be able to infer more equality predicates.

Reply:
> There is `break` to return the inner loop. I think for join conditions: `on(a.id = b.id and a.id = b.id2) where b.id = 10`, we should be able to infer more equality predicates.

It should be common optimization, `infer conditon`, Many rule about `join` need to it.